### PR TITLE
✨ Added graphql-ws

### LINKED ISSUE #288: Does Zues support the GraphQL over WebSocket Protocol?
I am using [Hasura](https://hasura.io) to manage my graphQL. As per [their docs](https://hasura.io/docs/latest/graphql/core/databases/postgres/subscriptions/index/) Hasura uses the [GraphQL over WebSocket Protocol](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) made by the Apollo people. 

I actually had no idea that there was a special protocol for this but it makes sense. After I tried to make a subscription with Zeus and it did not work I looked under the hood in index.ts and saw found code:

```ts
export const apiSubscription = (options: chainOptions) => (query: string) => {
  try {
    const queryString = options[0] + '?query=' + encodeURIComponent(query);
    const wsString = queryString.replace('http', 'ws');
    const host = (options.length > 1 && options[1]?.websocket?.[0]) || wsString;
    const webSocketOptions = options[1]?.websocket || [host];
    const ws = new WebSocket(...webSocketOptions);
    return {
      ws,
      on: (e: (args: any) => void) => {
        ws.onmessage = (event: any) => {
          if (event.data) {
            const parsed = JSON.parse(event.data);
            const data = parsed.data;
            return e(data);
          }
        };
      },
      off: (e: (args: any) => void) => {
        ws.onclose = e;
      },
      error: (e: (args: any) => void) => {
        ws.onerror = e;
      },
      open: (e: () => void) => {
        ws.onopen = e;
      },
    };
  } catch {
    throw new Error('No websockets implemented');
  }
};
```

So if I understand this correctly you basically make a Subscription() with a link and then a query just like when using Chain but then Zeus will generate a string with the query as a qeury on the url and create the websocket connection. 
What I didn't realize because it was not in the docs was that the Subscription function returns this object with .on .open .error .off and  also the ws object. So this leads me to assume that I am supposed to implement my own protocol with this. 

Is that indeed the idea? Is the idea that people implement their own protocol? If so has anyone implemented the GraphQL over WebSocket Protocol with Zues? Or am I missing something? I saw the --appollo flag but to me it did not seem to include an GraphQL over WebSocket Protocol implementation.

If this is the idea is it even a good idea to put the subscription in the query string? Because on the one hand this project does not make assumptions about the subscription protocol and you are supposed to take care of it yourself but then also it does make the assumption that you want your query in the the url query parameter. But maybe that is how a lot of people do it I don't know.

Also I want to suggest putting the error at the end if it was catchted into the new Error because the reason that it failed might not be that there is no websockets implemented.

### PR DESCRIPTION
**Pull request by GauBen**
Working to fix #288

- [x] Added [graphql-ws](https://github.com/enisdenjo/graphql-ws)
- [x] Preserve the current behavior (named *legacy*, because non-standard)
- [x] Implement `open` and `off`, if they still make sense
- [x] Write the documentation
- [x] Add graphql-ws as an optional peer dependency

### CODE REVIEW

**Comment by LMaxence** [comment]
Hello @aexol, could you update us on the status of this please ? There are no review, no feedbacks, and it is hard for us to know if we can expect this feature that we really really really need merged at some point or if we have to fork the project on our side.

**Comment by aexol** [comment]
Sorry I thought it is not completed. Why not make a plugin instead of forcing graphql-ws to the main code?

**Comment by GauBen** [comment]
It looks like Zeus support subscriptions out of the box but that's not the case because the implementation of [apiSubscription](https://github.com/graphql-editor/graphql-zeus/blob/master/src/TreeToTS/functions/new/apiSubscription.ts#L3) is non standard. This PR is more of a fix to add the standard implementation to the core, as a non-default feature flag to keep the current behavior, rather a new feature. This PR will make Zeus compatible with Apollo and Yoga subscriptions that both follow [this standard](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md).

I haven't written the documentation yet because I wanted to make sure that you were okay with changes to the core. If you agree on this design, I'll complete the todo list.

Also, I chose to implement it as a non-breaking change; both the current implementation and graphql-ws currently coexist in generated.ts: https://github.com/graphql-editor/graphql-zeus/pull/301/files#diff-96a2264dfde03a31fbad8f2aac04c691dd6a7291919fbcbf258ba30782f394daR664

Then, you may or may not make graphql-ws the default in the next major version.

**Comment by GauBen** [comment]
> If it won't break I am ok go ahead and finish it. 

Great news! I'll try to have it ready by the end of the week.

> you are implementing ts with strings

Not really, I did it the same way as you:
- Created a TS file in `src/TreeToTS/functions/apiSubscription/graphql-ws.ts`
- Added a bit of magic to `src/CLI/libBuilder.ts` to handle the new files

**Comment by GauBen** [comment]
Ready for review!

I feel like a lot of things are wrong:

### The type `SubscriptionToGraphQL` is not correctly implemented

In the current implementation, this is the `open` function:
```ts
return {
    // ...
	open: (e: () => void) => {
		ws.onopen = e;
	}
}
```

[link](https://github.com/graphql-editor/graphql-zeus/blob/59d55d8d0d71b315b97a01f4a1591b26d305727c/src/TreeToTS/functions/new/apiSubscription.ts#L27-L29)

And this is the signature:

```ts
export type SubscriptionToGraphQL<Z, T, SCLR extends ScalarDefinition> = {
  // ...
  open: () => void;
};
```

[link](https://github.com/graphql-editor/graphql-zeus/blob/59d55d8d0d71b315b97a01f4a1591b26d305727c/src/TreeToTS/functions/new/types.ts#L75)

### I committed documentation artifacts

Is this ok?

### There is no way to properly cleanup

I implemented the cleanup function with a Proxy over `ws.close`, but it feels dirty. Graphql-ws offers a proper way to unsubscribe and clean up:

https://github.com/enisdenjo/graphql-ws/blob/77682cd3d9bac748b5b24e2f8cf88e1d98c7a448/src/client.ts#L434-L442

### We should abstract the underlying transport

There are two competing standards, graphql-ws and graphql-sse, who share the exact same interface. We might be interested in moving from concrete APIs (`ws`, `open`...) to more abstract ones (`transport`, `cleanup`...), in order to make Zeus implementation-agnostic.

**Comment by aexol** [comment]
Thanks @GauBen I'll try to review next days

**Comment by GauBen** [comment]
Thank you! I'll sort the merge commits out soon.

File: package.json [hunk 1]
```diff
@@ -66,6 +67,14 @@
     "graphql-js-tree": "^0.1.1",
     "yargs": "^16.1.1"
   },
+  "peerDependencies": {
+    "graphql-wq": ">=5"
```
Comment (aexol) [inline package.json:71]: should be graphql-ws

File: src/TreeToTS/functions/apiSubscription/graphql-ws.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import { chainOptions } from '@/TreeToTS/functions/new/models';
+import { createClient, type Sink } from 'graphql-ws'; // keep
```
Comment (aexol) [inline src/TreeToTS/functions/apiSubscription/graphql-ws.ts:2]: ok

File: src/TreeToTS/index.ts [hunk 3]
```diff
@@ -29,6 +29,15 @@ export interface OperationDetails {
   operations: string[];
 }
 
+export interface ResolveOptions {
+  tree: ParserTree;
+  env?: Environment;
+  host?: string;
+  headers?: Record<string, string>;
+  esModule?: boolean;
+  subscriptions: 'legacy' | 'graphql-ws';
```
Comment (aexol) [inline src/TreeToTS/index.ts:38]: just make optional with default value for better backwards compatibility

File: src/TreeToTS/index.ts [hunk 4]
```diff
@@ -81,22 +92,11 @@ export class TreeToTS {
...
-    esModule,
-    headers,
-  }: {
-    tree: ParserTree;
-    env?: Environment;
-    host?: string;
-    headers?: Record<string, string>;
-    esModule?: boolean;
-  }) {
+  static resolveTreeSplit({ tree, env = 'browser', host, esModule, headers, subscriptions }: ResolveOptions) {
```
Comment (aexol) [inline src/TreeToTS/index.ts:99]: provide a default 'legacy' value

**Review by aexol — CHANGES_REQUESTED** [review]
Looks nice! Please fix those small errors and suggestions, so we can merge and release next version 🚀

**Comment by GauBen** [comment]
🎉

**Comment by aexol** [comment]
Please test `5.1.7` when ready :)

**Comment by aexol** [comment]
Thank you!

**Comment by GauBen** [comment]
Everything works on my side! 🎉

**Pull request merged**